contracts/Paper.sol

```diff
@@ -0,0 +1,156 @@
+        transferOwnership(daoAddress);
+        dopeWarsLootContract = IERC721Enumerable(dopeWarsLootContractAddress);
+        _mint(daoAddress, supply / 2);
+    }
+
+    function snapshot() public onlyOwner {
+        _snapshot();
+    }
+
+    // The following functions are overrides required by Solidity.
+
+    function _beforeTokenTransfer(address from, address to, uint256 amount)
+        internal
+        override(ERC20, ERC20Snapshot)
+    {
+        super._beforeTokenTransfer(from, to, amount);
+    }
+
+    /// @notice Allows the DAO to set a new contract address for Dope Wars Loot. This is
+    /// relevant in the event that DWL migrates to a new contract.
+    /// @param dopeWarsLootContractAddress_ The new contract address for DWL
+    function daoSetLootContractAddress(address dopeWarsLootContractAddress_)
+        external
+        onlyOwner
+    {
+        dopeWarsLootContractAddress = dopeWarsLootContractAddress_;
+        dopeWarsLootContract = IERC721Enumerable(dopeWarsLootContractAddress);
+    }
+
+    /// @notice Claim Paper for a given Dope Wars Loot ID
+    /// @param tokenId The tokenId of the Dope Wars Loot NFT
+    function claimById(uint256 tokenId) external {
+        // Follow the Checks-Effects-Interactions pattern to prevent reentrancy
+        // attacks
+
+        // Checks
+
+        // Check that the msgSender owns the token that is being claimed
+        require(
+            _msgSender() == dopeWarsLootContract.ownerOf(tokenId),
+            "MUST_OWN_TOKEN_ID"
+        );
+
+        // Further Checks, Effects, and Interactions are contained within the
+        // _claim() function
+        _claim(tokenId, _msgSender());
+    }
+
+    /// @notice Claim Paper for all tokens owned by the sender
+    /// @notice This function will run out of gas if you have too much loot! If
+    /// this is a concern, you should use claimRangeForOwner and claim Dope in
+    /// batches.
+    function claimAllForOwner() external {
+        uint256 tokenBalanceOwner = dopeWarsLootContract.balanceOf(_msgSender());
+
+        // Checks
+        require(tokenBalanceOwner > 0, "NO_TOKENS_OWNED");
+
+        // i < tokenBalanceOwner because tokenBalanceOwner is 1-indexed
+        for (uint256 i = 0; i < tokenBalanceOwner; i++) {
+            // Further Checks, Effects, and Interactions are contained within
+            // the _claim() function
+            _claim(
+                dopeWarsLootContract.tokenOfOwnerByIndex(_msgSender(), i),
+                _msgSender()
+            );
+        }
+    }
+
+    /// @notice Claim Paper for all tokens owned by the sender within a
+    /// given range
+    /// @notice This function is useful if you own too much DWL to claim all at
+    /// once or if you want to leave some Paper unclaimed. If you leave Paper
+    /// unclaimed, however, you cannot claim it once the next season starts.
+    function claimRangeForOwner(uint256 ownerIndexStart, uint256 ownerIndexEnd)
+        external
```

Comment:
```suggestion
        external
        nonReentrant
```

Reply:
Don't think you need to use ReentrancyGuard for those